### PR TITLE
feat: stabilize credential-process and registry-auth

### LINKED ISSUE #8933: Tracking Issue for credential-process RFC 2730
**Summary**

RFC: https://github.com/rust-lang/rfcs/pull/2730
Implementation: #8934
Documentation: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#credential-process
Issues: https://github.com/rust-lang/cargo/labels/Z-credential-process

This feature provides a configuration option to specify a process to fetch a token to authenticate with a registry.

**Unresolved issues**

* [X] Is this approach useful enough?  Things like macos keychain don't protect against being executed to extract tokens (and I don't see a way to require a password, or force the process to be untrusted).  The lack of signatures also cause issues (each update of the toolchain will cause it to become untrusted again).
  * Making the built-in providers like `cargo:macos-keychain` part of the cargo binary improves security here, since only the Cargo binary is accessing the Keychain. If we get Cargo signed by Apple, then it would be further improved.
* [X] Should the `login` API be changed? Providers that need to be interactive need to be able to read from stdin.
  * This is worked around by having the credential-provider re-open `stdin` from `/dev/tty` or `$CONIN`
* [X] Should the storage key off the registry name? [RFC 3139](https://github.com/rust-lang/rfcs/pull/3139) discusses problems with this. RESOLVED: credential helpers are now keying off the index url via #10592.
* [x] Should it be possible to pass additional arguments on the command line for `cargo login` that would be sent to the provider? E.g. `cargo login -- --extra-arg-for-provider` 
  * Implemented in #12499
* [X] Figure out the name for `cargo:basic`: `cargo:token-from-stdout`
* [X] Remove cargo-1password as a built-in (add the `fn main` so it can be published separately).
* [x] Create a new `-Z` flag for `cargo:paseto`.
  * [x] Publish `cargo-credential-1password` to crates.io
* [x] Figure out MSRV support. rust-lang/cargo#12381 is proposing to set MSRV for all packages in the repo. I think it will be important that the credential provider libraries have a relatively relaxed requirement (like stable-2 or something). I think this will be important to figure out before stabilization. This may be complicated with CI costs, so I'm wondering if `cargo-credential` could grow a few independent tests on its own so we could test just the library with an older rust? https://github.com/rust-lang/cargo/pull/12623
* [x] Testing of forwards-compatibility support. It is important that when a user installs a new credential provider that it does not necessarily cause older `cargo`s to fail. There is a version field when we need to do breaking changes, but we need tests that are actually verifying that is working. https://github.com/rust-lang/cargo/pull/12622
  * [x] A test with some manually generated JSON with an invalid version file fed into a credential provider should provide the correct error message.
  * [x] A test with some manually generated JSON with the current V1 format that is fed into a credential provider, and validate that it behaves correctly. This will make sure that if any changes to the credential provider are made that are not compatible with the old format that the test will fail.
* [x] Resolve rust-lang/cargo#12642
* [ ] Stabilization PR - #12649
* [ ] Publish `cargo-credential` `v0.4` to [crates.io](https://crates.io/crates/cargo-credential)
* [ ] Publish `cargo-credential-1password` `v0.4` to [crates.io](https://crates.io/crates/cargo-credential-1password)

**About tracking issues**

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### LINKED ISSUE #10474: Tracking Issue for alternative registry authentication (RFC 3139)
### Summary

RFC: [#3139](https://github.com/rust-lang/rfcs/pull/3139) 
Implementation: https://github.com/rust-lang/cargo/pull/10592
Documentation: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#registry-auth
Issue: https://github.com/rust-lang/cargo/labels/Z-registry-auth

This feature adds the ability to authenticate additional endpoints to a registry, including downloading crates.

### Unresolved Issues

* [x] Do registries need a more fine-grained switch for which API commands require authentication?
* [X] The RFC mentions adding `--token` to additional commands like `install` and `search`, but we are leaning away from allowing tokens from being passed in on the command-line due to the ease of leaking. Should the `--token` flag be added or no?  `--token` won't be added for now.
* [X] Consider changing the name and form of the `X-` header. See https://github.com/rust-lang/rfcs/pull/3139#discussion_r819834286 and https://github.com/rust-lang/rfcs/pull/3139#discussion_r823436983  Cargo now uses the www-authenticate header with the `Cargo` scheme and the `login_url` value, as in `WWW-Authenticate: Cargo login_url="https://test-registry-login/me`.
* [X] Will there be any concerns with the interaction with [RFC 3231](https://github.com/rust-lang/rfcs/pull/3231) (asymmetric tokens)?
* [ ] Require a `credential-provider` to be defined in order to use authenticated registries

Stabilization tracked in #8933

### Future Extensions

* Support authentication with git indexes.  Preferably, cargo will transition to [HTTP indexes](https://github.com/rust-lang/cargo/issues/9069) which will make this not necessary.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### PR DESCRIPTION
**Pull request by arlosi**
Stabilization PR for `registry-auth` and `credential-process`.

Tracking approval of this stabilization is done in the via the FCP in [#8933](https://github.com/rust-lang/cargo/issues/8933#issuecomment-1711990123). This PR is here to help reviewers of the FCP.

* Stabilizes `registry-auth` and `credential-process`
* Makes authenticated registries require a credential provider
* Adds stable documentation for credential providers and authenticated registries

Closes #8933
Closes #10474

### CODE REVIEW

File: src/doc/src/reference/registry-authentication.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+* `CARGO` --- Path to the `cargo` binary executing the command.
+* `CARGO_REGISTRY_INDEX_URL` --- The URL of the registry index.
+* `CARGO_REGISTRY_NAME_OPT` --- Optional name of the registry. Should not be used as a lookup key.
+
+[`cargo login`]: ../commands/cargo-login.md
+[`cargo logout`]: ../commands/cargo-logout.md
+
+## External providers
+For external providers that follow Cargo's [credential provider protocol](credential-provider-protocol.md)
+use the path to the executable (or the executable name if on the `PATH`).
```
Comment (ehuss) [inline src/doc/src/reference/registry-authentication.md:92]: This sentence was a little confusing on my first read. Does something like the following seem a little clearer?

```suggestion
For external providers that follow Cargo's [credential provider protocol](credential-provider-protocol.md),
the configuration value should be a string with the path to the executable (or the executable name if on the `PATH`).
```

**Comment by ehuss** [comment]
Thanks so much for writing all the documentation, it looks great!

Just some minor notes about the documentation:

Can you update https://github.com/rust-lang/cargo/blob/HEAD/src/doc/src/reference/environment-variables.md to include `CARGO_CREDENTIAL_ALIAS_<name>` and `CARGO_REGISTRIES_<name>_CREDENTIAL_PROVIDER` and `CARGO_REGISTRY_GLOBAL_CREDENTIAL_PROVIDERS`?

----

One thing I'd like to chat about is documenting the stability of the built-in credential providers. I'm concerned that we may not be able to support them indefinitely, at which point we would need to remove one which would break cargo's general compatibility guarantees. For example, let's say Apple or Microsoft suddenly start requiring signed binaries in order to access the global credential stores, and we can't accommodate that. Or, let's say libsecret stops development and distros stop using it. I'm wondering it there should be a note in the auth or config chapter about that? I don't want a note that sounds scary, since it is unlikely that we would stop supporting them, and we would also try to provide a warning ahead of time if possible if we plan to remove it. But I do think it is useful to clearly communicate to users what our expectations around compatibility are.

An alternate idea would be to have a separate "compatibility contract" that would generally state how cargo treats backwards and forwards compatibility, and then exceptions like this (and other things like RUSTFLAGS or platform support) could be mentioned there. Right now, we only have https://doc.crates.io/contrib/design.html which spells that out, but that is not end-user documentation. Perhaps that could move to the cargo guide?

Any thoughts on that?

File: src/doc/src/reference/registry-authentication.md (hunk 1 above)
Comment (arlosi) [reply src/doc/src/reference/registry-authentication.md]: Yes, this is better. Thanks!

**Comment by arlosi** [comment]
Good catch on the missing docs in `environment-variables.md`. Updated.

> I'm concerned that we may not be able to support them indefinitely, at which point we would need to remove one which would break cargo's general compatibility guarantees.

I've added a note in `registry-authentication.md` that says:
> Cargo includes several built-in credential providers. The available built-in providers may change in future Cargo releases.

Do you think that's sufficient?

**Comment by ehuss** [comment]
> Do you think that's sufficient?

Seems good to me.

**Comment by Eh2406** [comment]
> Cargo includes several built-in credential providers. The available built-in providers may change in future Cargo releases. Although, we do not have plans to do so in the near future. If plans change we will update this documentation.

Or sum such.

**Comment by ehuss** [comment]
I forgot, can you also update [`publish.py`](https://github.com/rust-lang/cargo/blob/master/publish.py) so that it includes all of the credential crates? 1password doesn't *need* to be included, but I don't think it will cause any harm to include it.

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Your efforts on this are much appreciated, Arlo! Also thanks everyone for reviewing and testing on this feature :)

We're not going to wait for the FCP ended, as the beta branching-off window is approaching, and the FCP [was started 9 days ago](https://github.com/rust-lang/cargo/issues/8933#issuecomment-1711990145). If you encounter any issue with this stabilization, please report it here or open a new issue.

@bors r+

**Pull request merged**